R/chunk_req_perform.R

```diff
@@ -0,0 +1,204 @@
+#' Chunk a request
+#'
+#' @inheritParams req_perform
+#' @param chunk_size The size of each chunk.
+#' @param data The data to chunk.
+#' @param apply_chunk A function that applies the chunk to the request. It
+#'   takes two arguments:
+#'
+#'   1. `req`: the original request.
+#'   2. `chunk`: the current data chunk.#'
+#' @return For `req_chunk()` a list of requests. For `chunk_req_perform()` a
+#'   list of parsed responses.
+#' @export
+#'
+#' @examples
+#' req <- request("https://api.restful-api.dev/objects")
+#'
+#' ids <- 1:7
+#' chunk_size <- 3
+#'
+#' apply_chunk <- function(req, chunk) {
+#'   chunk <- rlang::set_names(chunk, "id")
+#'   req_url_query(req, !!!chunk)
+#' }
+#'
+#' parse_resp <- function(resp) {
+#'   parsed <- resp_body_json(resp)
+#'   data.frame(
+#'     id = sapply(parsed, function(x) x$id),
+#'     name = sapply(parsed, function(x) x$name)
+#'   )
+#' }
+#'
+#' # in most cases `chunk_req_perform()` should be sufficient
+#' \dontrun{
+#' responses <- chunk_req_perform(
+#'   req,
+#'   chunk_size = 10,
+#'   data = ids,
+#'   apply_chunk = apply_chunk,
+#'   parse_resp = parse_resp
+#' )
+#' }
+#'
+#' # in case you need more control of how the requests are performed you can create
+#' # the requests first with `req_chunk()` ...
+#' requests <- req_chunk(
+#'   req,
+#'   chunk_size = 3,
+#'   data = ids,
+#'   apply_chunk = apply_chunk
+#' )
+#'
+#' # a simple list of requests
+#' requests
+#'
+#' # ... which you can perform with `req_perform()`
+#' \dontrun{
+#' n <- length(requests)
+#' paths <- character(n)
+#' responses <- list()
+#' for (i in seq2(1, n)) {
+#'   paths[[i]] <- tempfile()
+#'   responses[[i]] <- req_perform(requests[[i]], path = paths[[i]])
+#' }
+#' }
+req_chunk <- function(req,
+                      data,
+                      chunk_size,
+                      apply_chunk) {
+  check_request(req)
+  chunks <- vec_chop_by_size(data, chunk_size)
+  check_function2(apply_chunk, args = c("req", "chunk"))
+
+  n <- length(chunks)
+  requests <- lapply(chunks, function(chunk) apply_chunk(req, chunk))
+  new_chunked_request(requests, chunks)
```

Comment:
This is different to all the other `req_` functions in that it now generates a list of requests. I think we need to either do the chunking lazily, not export this function, or give it a different prefix.

Reply:
I now also implemented this as a policy. This simplifies the interface of `chunk_req_perform()` and makes the whole API more similar to the pagination API.